runelite-client/src/main/java/net/runelite/client/plugins/inferno/InfernoNPC.java

```diff
@@ -25,162 +25,472 @@
+				boolean hasRange = this.getType().getDefaultAttack() == Attack.MELEE ? bestNexArea.isInMeleeDistance(target)
+					: bestNexArea.distanceTo(target) <= this.getType().getRange();
+
+				if (hasLos && hasRange)
+				{
+					//System.out.println("Move location has LOS, can walk to: true");
+					safeSpotCache.put(bestNextLocation, 1);
+					return true;
+				}
+
+				currentPosition = bestNextLocation;
+			}
+		}
+	}
+
+	boolean couldAttackPrevTick(Client client, WorldPoint lastPlayerLocation)
+	{
+		return new WorldArea(lastPlayerLocation, 1, 1).hasLineOfSightTo(client, this.getNpc().getWorldArea());
+	}
+
+	void gameTick(Client client, WorldPoint lastPlayerLocation, boolean finalPhase)
+	{
+		safeSpotCache.clear();
+
+		if (ticksTillNextAttack > 0)
+		{
+			this.ticksTillNextAttack--;
+		}
+
+		//Jad animation detection
+		if (this.getType() == Type.JAD && this.getNpc().getAnimation() != -1 && this.getNpc().getAnimation() != this.lastAnimation)
+		{
+			final InfernoNPC.Attack currentAttack = InfernoNPC.Attack.attackFromId(this.getNpc().getAnimation());
+
+			if (currentAttack != null && currentAttack != Attack.UNKNOWN)
+			{
+				this.updateNextAttack(currentAttack, this.getType().getTicksAfterAnimation());
+			}
+		}
+
+		if (ticksTillNextAttack <= 0)
+		{
+			if (this.getType() == Type.ZUK)
+			{
+				if (this.getNpc().getAnimation() == AnimationID.TZKAL_ZUK)
+				{
+					if (finalPhase)
+					{
+						this.updateNextAttack(this.getType().getDefaultAttack(), 7);
+					}
+					else
+					{
+						this.updateNextAttack(this.getType().getDefaultAttack(), 10);
+					}
+				}
+			}
+			else if (this.getType() == Type.JAD)
+			{
+				if (this.getNextAttack() != Attack.UNKNOWN)
+				{
+					// Jad's cycle continuous after his animation + attack but there's no animation to alert it
+					this.updateNextAttack(this.getType().getDefaultAttack(), 8);
+				}
+			}
+			else if (this.getType() == Type.BLOB)
+			{
+				//RS pathfinding + LOS = hell, so if it can attack you the tick you were on previously, start attack cycle
+				if (!this.lastCanAttack && this.couldAttackPrevTick(client, lastPlayerLocation))
+				{
+					this.updateNextAttack(Attack.UNKNOWN, 3);
+				}
+				//If there's no animation when coming out of the safespot, the blob is detecting prayer
+				else if (!this.lastCanAttack && this.canAttack(client, client.getLocalPlayer().getWorldLocation()))
+				{
+					this.updateNextAttack(Attack.UNKNOWN, 4);
+				}
+				//This will activate another attack cycle
+				else if (this.getNpc().getAnimation() != -1)
+				{
+					this.updateNextAttack(this.getType().getDefaultAttack(), this.getType().getTicksAfterAnimation());
+				}
+			}
+			// Range + LOS check for bat because it suffers from the defense animation bug, also dont activate on "stand" animation
+			else if (this.getType() == Type.BAT)
+			{
+				if (this.canAttack(client, client.getLocalPlayer().getWorldLocation()) && this.getNpc().getAnimation() != 7577
```

Comment:
AnimationID

Reply:
This animation does not exist in AnimationID

Reply:
AnimationID isn't computer generated, just add htem all there